sbncode/LArRecoProducer/CRUMBS_module.cc

```diff
@@ -0,0 +1,681 @@
+	const std::vector<art::Ptr<anab::T0> > sliceCRTHitT0s = this->GetCRTHitT0s(e, slice, handlePFPs, handleSlices);
+
+	this->FillCRTVars(sliceCRTTrackT0s, sliceCRTHitT0s);
+
+	const art::Ptr<larpandoraobj::PFParticleMetadata> pfpMeta = pfpMetaVec.front();
+	std::map<std::string, float> propertiesMap = pfpMeta->GetPropertiesMap();
+      
+	this->FillPandoraNuScoreVars(propertiesMap);
+
+	tpc_StoppingChi2CosmicRatio = this->GetLongestTrackStoppingChi2Ratio(e, slice, handlePFPs, handleSlices);
+    
+	const art::Ptr<sbn::SimpleFlashMatch> flashmatch = pfpFMVec.front();
+	pds_FMTotalScore = flashmatch->score.total;
+	pds_FMPE = flashmatch->light.pe;
+	pds_FMTime = flashmatch->time;
+      
+	if(pds_FMTime == -9999)
+	  pds_FMTime = -100;
+
+	const float score = fMVAReader->EvaluateMVA(fMVAName);
+
+	CRUMBSResult thisResult(score, tpc_CRFracHitsInLongestTrack, tpc_CRLongestTrackDeflection, tpc_CRLongestTrackDirY, tpc_CRNHitsMax,
+				tpc_NuEigenRatioInSphere, tpc_NuNFinalStatePfos, tpc_NuNHitsTotal, tpc_NuNSpacePointsInSphere, tpc_NuVertexY, 
+				tpc_NuWeightedDirZ, tpc_StoppingChi2CosmicRatio, pds_FMTotalScore, pds_FMPE, pds_FMTime, crt_TrackScore, crt_HitScore, 
+				crt_TrackTime, crt_HitTime);
+
+	resultsVec->push_back(thisResult);
+	util::CreateAssn(*this, e, *resultsVec, slice, *sliceAssns);
+
+	if(fTrainingMode)
+	  {
+	    std::vector<art::Ptr<recob::Hit> > sliceHits = this->GetAllSliceHits(e, slice, handleSlices);
+	    std::map<int, float> puritiesMap = this->SlicePurity(e, sliceHits);
+	    const int truthId = this->SliceTruthId(puritiesMap);
+
+	    slicePDG = primary->PdgCode();
+	    matchedType = fGenTypeMap[truthId];
+	    matchedPurity = puritiesMap[truthId];
+	    matchedCompleteness = this->SliceCompleteness(e, sliceHits, allHits, truthId);
+      
+	    fSliceTree->Fill();
+	  }
+      }
+
+    e.put(std::move(resultsVec));
+    e.put(std::move(sliceAssns));
+  }
+
+  void CRUMBS::FillCRTVars(std::vector<art::Ptr<anab::T0> > trackT0s, std::vector<art::Ptr<anab::T0> > hitT0s)
+  {
+    if (trackT0s.size() != 0){
+      crt_TrackScore = 999999;
+      for(auto const crttrackmatcht0 : trackT0s)
+	{
+	  if(crttrackmatcht0->TriggerConfidence() < crt_TrackScore)
+	    {
+	      crt_TrackScore = crttrackmatcht0->TriggerConfidence();
+	      crt_TrackTime = crttrackmatcht0->Time() * 1e-3;
+	    }
+	}
+    }
+  
+    if (hitT0s.size() != 0){
+      crt_HitScore = 999999;
+      for(auto const crthitmatcht0 : hitT0s)
+	{
+	  if(crthitmatcht0->TriggerConfidence() < crt_HitScore)
+	    {
+	      crt_HitScore = crthitmatcht0->TriggerConfidence();
+	      crt_HitTime = crthitmatcht0->Time() * 1e-3;
+	    }
+	}
+    }
+  }
+
+  void CRUMBS::FillPandoraNuScoreVars(std::map<std::string, float> propertiesMap)
+  {
+    auto propertiesMapIter = propertiesMap.find("NuScore");
+    if (propertiesMapIter == propertiesMap.end()){
+      std::cout << "Error finding variable -- NuScore" << std::endl;
```

Comment:
Preface these messages with `CRUMBS_module:` so that the poor user can find where they come from.

Personally I would `abort()` rather than simply `return` here. The user who intended to run CRUMBS is not getting what they wanted, and will probably be happier that you make sure they notice now, rather than digging through their grid logs to find out why things aren't filled right.